Race condition in Consul result backend when saving a result

I'm encountering an [issue](https://groups.google.com/forum/#!topic/consul-tool/U1GXnuNmMN8) using the Consul result backend which I suspect is NOT a Celery issue as such but rather [an issue in the lack of thread safety in the library the backend uses](https://github.com/cablehead/python-consul/issues/144).

The symptoms are consistent with confusion between the sorts of responses that Consul gives to the library when creating saved results, and expiring those saved results. So, if Celery both writes and expires saved results using threads and, the lack of thread-safety in the python-consul library could produce the results described.

Can somebody confirm if Celery access to a results backend involves multiple threads?

I finally have what I think is a smoking gun pointing to an issue in Celery. My setup is much as before though with versions as follows:

```
$ celery report

software -> celery:4.4.5 (cliffs) kombu:4.6.10 py:3.8.2
            billiard:3.6.3.0 py-amqp:2.6.0
platform -> system:Linux arch:64bit, ELF
            kernel version:5.4.0-37-generic imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled
```

and using, as before, the Consul result backend. I modified the Celery code slightly to emit these logs:

```
2020-06-24 09:16:03,336 [WARNING] ForkPoolWorker-27: Trying to fetch key celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05 from Consul
2020-06-24 09:16:03,336 [WARNING] ForkPoolWorker-27: request get="celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05"
2020-06-24 09:16:09,033 [WARNING] ForkPoolWorker-27: response data="{'ID': '7abf2295-6dca-6f7d-ed4c-4bff4b60955f'}", decode=Value is_id=False
2020-06-24 09:16:09,033 [ERROR] ForkPoolWorker-27: cannot get celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05: 'str' object has no attribute 'get' (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 64, in get
    _, data = self.client.kv.get(key)
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 553, in get
    return self.agent.http.get(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py", line 21, in get
    return callback(self.response(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 232, in cb
    if item.get(decode) is not None:
AttributeError: 'str' object has no attribute 'get'
2020-06-24 09:16:09,035 [WARNING] ForkPoolWorker-27: /usr/local/lib/python3.8/dist-packages/celery/app/trace.py:622: RuntimeWarning: Exception raised outside body: AttributeError("'str' object has no attribute 'get'"):
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/celery/app/trace.py", line 479, in trace_task
    mark_as_done(
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 158, in mark_as_done
    self.store_result(task_id, result, state, request=request)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 442, in store_result
    self._store_result(task_id, result, state, traceback,
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 853, in _store_result
    current_meta = self._get_task_meta_for(task_id)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 871, in _get_task_meta_for
    meta = self.get(self.get_key_for_task(task_id))
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 64, in get
    _, data = self.client.kv.get(key)
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 553, in get
    return self.agent.http.get(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py", line 21, in get
    return callback(self.response(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 232, in cb
    if item.get(decode) is not None:
AttributeError: 'str' object has no attribute 'get'

  warn(RuntimeWarning(
2020-06-24 09:16:09,036 [WARNING] ForkPoolWorker-27: Trying to fetch key celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05 from Consul
2020-06-24 09:16:09,036 [WARNING] ForkPoolWorker-27: request get="celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05"
2020-06-24 09:16:09,053 [WARNING] ForkPoolWorker-27: Trying to create Consul session celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05 with TTL 86400
2020-06-24 09:16:09,199 [WARNING] ForkPoolWorker-27: response data="True", decode=False is_id=True
2020-06-24 09:16:09,199 [ERROR] ForkPoolWorker-27: cannot save result for b'celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05'={"status": "FAILURE", "result": {"exc_type": "AttributeError", "exc_message": ["'str' object has no attribute 'get'"], "exc_module": "builtins"}, "traceback": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.8/dist-packages/celery/app/trace.py\", line 479, in trace_task\n    mark_as_done(\n  File \"/usr/local/lib/python3.8/dist-packages/celery/backends/base.py\", line 158, in mark_as_done\n    self.store_result(task_id, result, state, request=request)\n  File \"/usr/local/lib/python3.8/dist-packages/celery/backends/base.py\", line 442, in store_result\n    self._store_result(task_id, result, state, traceback,\n  File \"/usr/local/lib/python3.8/dist-packages/celery/backends/base.py\", line 853, in _store_result\n    current_meta = self._get_task_meta_for(task_id)\n  File \"/usr/local/lib/python3.8/dist-packages/celery/backends/base.py\", line 871, in _get_task_meta_for\n    meta = self.get(self.get_key_for_task(task_id))\n  File \"/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py\", line 64, in get\n    _, data = self.client.kv.get(key)\n  File \"/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py\", line 553, in get\n    return self.agent.http.get(\n  File \"/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py\", line 21, in get\n    return callback(self.response(\n  File \"/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py\", line 232, in cb\n    if item.get(decode) is not None:\nAttributeError: 'str' object has no attribute 'get'\n", "children": [[["b3cc0538-e39d-4ff5-accd-d1d0c0e3e30e", null], null], [["14b3b7a0-1374-4ade-adc6-827f55cbe7ae", null], null], [["8a156251-4545-4bf5-9351-200d79b2a757", null], null], [["2563d59b-7885-41e2-b0f2-e01a0fe173c9", null], null], [["840e501b-744b-4d87-9400-7416391dd05a", null], null], [["b08a15a5-5abb-4103-b4ed-8ed16a760521", null], null], [["eb2de10e-f49b-47b8-a2e1-71cbc983774c", null], null], [["4546e0f5-dd1b-45c6-bfba-6fd3de9cc5e4", null], null], [["cd5a9a00-fdae-4bc4-94a9-eb784c1d5be7", null], null], [["e1797005-0f2f-47c9-a1df-fdf26edc07cd", null], null], [["49b95dec-be25-4f86-a295-4fc4650c0665", null], null], [["96f37a25-9b76-4de5-98c3-4e504816933c", null], null], [["6ef42ea5-a45c-4f43-83e3-29c4964bf134", null], null], [["08efabc6-81e1-4ff7-9970-435d4efcbb2c", null], null], [["4ade8f72-13fb-433c-aec4-844f019607ff", null], null], [["11b1456c-f558-461b-a872-0044aa29c954", null], null], [["d8fb6ad9-9669-4c8a-8e73-aef0ffff452a", null], null], [["890a6831-e033-4b63-9199-2f4858b6dd40", null], null], [["f2d7b230-ac3c-489f-b3a3-b5956b379a2d", null], null], [["34d1a8a8-9f8f-44b5-88e3-1ee8713d7a8a", null], null], [["2522e3dc-7bf2-4cb8-90cb-99a9011844e0", null], null]], "date_done": "2020-06-24T09:16:09.036404", "parent_id": "a91ffd09-bff0-4d44-a954-e596c08de35c", "task_id": "601e8b28-9dcf-4e68-847b-5e1875a8eb05"}: 'bool' object is not subscriptable (FAILURE)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/celery/app/trace.py", line 479, in trace_task
    mark_as_done(
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 158, in mark_as_done
    self.store_result(task_id, result, state, request=request)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 442, in store_result
    self._store_result(task_id, result, state, traceback,
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 853, in _store_result
    current_meta = self._get_task_meta_for(task_id)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 871, in _get_task_meta_for
    meta = self.get(self.get_key_for_task(task_id))
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 64, in get
    _, data = self.client.kv.get(key)
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 553, in get
    return self.agent.http.get(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py", line 21, in get
    return callback(self.response(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 232, in cb
    if item.get(decode) is not None:
AttributeError: 'str' object has no attribute 'get'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 78, in set
    return self.set_real(key, value, state)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 99, in set_real
    session_id = self.client.session.create(name=session_name,
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 1763, in create
    return self.agent.http.put(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py", line 26, in put
    return callback(self.response(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 235, in cb
    data = data['ID']
TypeError: 'bool' object is not subscriptable
2020-06-24 09:16:09,200 [ERROR] ForkPoolWorker-27: Task paiyroll.tasks.viewflow_dequeue[601e8b28-9dcf-4e68-847b-5e1875a8eb05] raised unexpected: AttributeError("'str' object has no attribute 'get'")
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/celery/app/trace.py", line 479, in trace_task
    mark_as_done(
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 158, in mark_as_done
    self.store_result(task_id, result, state, request=request)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 442, in store_result
    self._store_result(task_id, result, state, traceback,
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 853, in _store_result
    current_meta = self._get_task_meta_for(task_id)
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/base.py", line 871, in _get_task_meta_for
    meta = self.get(self.get_key_for_task(task_id))
  File "/usr/local/lib/python3.8/dist-packages/celery/backends/consul.py", line 64, in get
    _, data = self.client.kv.get(key)
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 553, in get
    return self.agent.http.get(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/std.py", line 21, in get
    return callback(self.response(
  File "/usr/local/lib/python3.8/dist-packages/python_consul-1.1.0-py3.8.egg/consul/base.py", line 232, in cb
    if item.get(decode) is not None:
AttributeError: 'str' object has no attribute 'get'
```

However, it is probably easier to see what is going on by looking at the attached wireshark packet capture.
[wireshark.pcapng.gz](https://github.com/celery/celery/files/4824820/wireshark.pcapng.gz):

1. Use gunzip to restore the file.
1. Load into wireshark.
1. Filter the display by "http".
1. Order by time column.
1. Look at the numbered entries (i.e. the first column) 8953 and 8963.
1. Note that the log timestamps and wireshark timestamps are 1 hour apart, so 09:00 is the log is 10:00 in wireshark.

Entries 8953 and 8963 seem to be a request response pair between Consul on port 8500 and the instance of Celery running on port 50268 (which I take to be ForkPoolWorker-27). Note the request is:

```
GET http://localhost:8500/v1/kv/celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05?consistent=1
```

and the response is encoded as application/json with content:

```
true
```

Note that `true` is *not* a valid response to the `GET`. It is actually the (valid) response to the earlier request in entry 8940:

```
PUT /v1/kv/celery-task-meta-4546e0f5-dd1b-45c6-bfba-6fd3de9cc5e4?acquire=128ced02-8f69-5327-69f6-56cbec3b3847 HTTP/1.1\r\n
```

which successfully stored the result of an earlier task.

The obvious implication is that after initiating the storing of the result for celery-task-meta-4546e0f5-dd1b-45c6-bfba-6fd3de9cc5e4, Celery did not wait for the acknowledgement `true`. Then when celery-task-meta-601e8b28-9dcf-4e68-847b-5e1875a8eb05 came along, it ended up using HTTP 1.1 pipelining, and mistaking the earlier response for its own.

Does that make sense?

--- a/celery/backends/consul.py
+++ b/celery/backends/consul.py
@@ -31,7 +31,6 @@ class ConsulBackend(KeyValueStoreBackend):
 
     supports_autoexpire = True
 
-    client = None
     consistency = 'consistent'
     path = None
 
@@ -40,15 +39,33 @@ class ConsulBackend(KeyValueStoreBackend):
 
         if self.consul is None:
             raise ImproperlyConfigured(CONSUL_MISSING)
-
+        #
+        # By default, for correctness, we use a client connection per
+        # operation. If set, self.one_client will be used for all operations.
+        # This provides for the original behaviour to be selected, and is
+        # also convenient for mocking in the unit tests.
+        #
+        self.one_client = None
         self._init_from_params(**parse_url(self.url))
 
     def _init_from_params(self, hostname, port, virtual_host, **params):
         logger.debug('Setting on Consul client to connect to %s:%d',
                      hostname, port)
         self.path = virtual_host
-        self.client = consul.Consul(host=hostname, port=port,
-                                    consistency=self.consistency)
+        self.hostname = hostname
+        self.port = port
+        #
+        # Optionally, allow a single client connection to be used to reduce
+        # the connection load on Consul by adding a "one_client=1" parameter
+        # to the URL.
+        #
+        if params.get('one_client', None):
+            self.one_client = self.client()
+
+    def client(self):
+        return self.one_client or consul.Consul(host=self.hostname,
+                                                port=self.port,
+                                                consistency=self.consistency)
 
     def _key_to_consul_key(self, key):
         key = bytes_to_str(key)
@@ -58,7 +75,7 @@ class ConsulBackend(KeyValueStoreBackend):
         key = self._key_to_consul_key(key)
         logger.debug('Trying to fetch key %s from Consul', key)
         try:
-            _, data = self.client.kv.get(key)
+            _, data = self.client().kv.get(key)
             return data['Value']
         except TypeError:
             pass
@@ -84,17 +101,16 @@ class ConsulBackend(KeyValueStoreBackend):
 
         logger.debug('Trying to create Consul session %s with TTL %d',
                      session_name, self.expires)
-        session_id = self.client.session.create(name=session_name,
-                                                behavior='delete',
-                                                ttl=self.expires)
+        client = self.client()
+        session_id = client.session.create(name=session_name,
+                                           behavior='delete',
+                                           ttl=self.expires)
         logger.debug('Created Consul session %s', session_id)
 
         logger.debug('Writing key %s to Consul', key)
-        return self.client.kv.put(key=key,
-                                  value=value,
-                                  acquire=session_id)
+        return client.kv.put(key=key, value=value, acquire=session_id)
 
     def delete(self, key):
         key = self._key_to_consul_key(key)
         logger.debug('Removing key %s from Consul', key)
-        return self.client.kv.delete(key)
+        return self.client().kv.delete(key)

--- a/docs/userguide/configuration.rst
+++ b/docs/userguide/configuration.rst
@@ -2016,14 +2016,52 @@ without any further configuration. For larger clusters you could use NFS,
 Consul K/V store backend settings
 ---------------------------------
 
-The Consul backend can be configured using a URL, for example:
+.. note::
+
+    The Consul backend requires the :pypi:`python-consul2` library:
+
+    To install this package use :command:`pip`:
+
+    .. code-block:: console
+
+        $ pip install python-consul2
+
+The Consul backend can be configured using a URL, for example::
 
     CELERY_RESULT_BACKEND = 'consul://localhost:8500/'
 
-The backend will storage results in the K/V store of Consul
-as individual keys.
+or::
 
-The backend supports auto expire of results using TTLs in Consul.
+    result_backend = 'consul://localhost:8500/'
+
+The backend will store results in the K/V store of Consul
+as individual keys. The backend supports auto expire of results using TTLs in
+Consul. The full syntax of the URL is::
+
+    consul://host:port[?one_client=1]
+
+The URL is formed out of the following parts:
+
+* ``host``
+
+    Host name of the Consul server.
+
+* ``port``
+
+    The port the Consul server is listening to.
+
+* ``one_client``
+
+    By default, for correctness, the backend uses a separate client connection
+    per operation. In cases of extreme load, the rate of creation of new
+    connections can cause HTTP 429 "too many connections" error responses from
+    the Consul server when under load. The recommended way to handle this is to
+    enable retries in ``python-consul2`` using the patch at
+    https://github.com/poppyred/python-consul2/pull/31.
+
+    Alternatively, if ``one_client`` is set, a single client connection will be
+    used for all operations instead. This should eliminate the HTTP 429 errors,
+    but the storage of results in the backend can become unreliable.
 
 .. _conf-messaging:
 

--- a/t/unit/backends/test_consul.py
+++ b/t/unit/backends/test_consul.py
@@ -22,8 +22,8 @@ class test_ConsulBackend:
     def test_get(self):
         index = 100
         data = {'Key': 'test-consul-1', 'Value': 'mypayload'}
-        self.backend.client = Mock(name='c.client')
-        self.backend.client.kv.get.return_value = (index, data)
+        self.backend.one_client = Mock(name='c.client')
+        self.backend.one_client.kv.get.return_value = (index, data)
         assert self.backend.get(data['Key']) == 'mypayload'
 
     def test_index_bytes_key(self):

--- a/t/unit/backends/test_consul.py
+++ b/t/unit/backends/test_consul.py
@@ -26,6 +26,17 @@ class test_ConsulBackend:
         self.backend.one_client.kv.get.return_value = (index, data)
         assert self.backend.get(data['Key']) == 'mypayload'
 
+    def test_set(self):
+        self.backend.one_client = Mock(name='c.client')
+        self.backend.one_client.session.create.return_value = 'c8dfa770-4ea3-2ee9-d141-98cf0bfe9c59'
+        self.backend.one_client.kv.put.return_value = True
+        assert self.backend.set('Key', 'Value') is True
+
+    def test_delete(self):
+        self.backend.one_client = Mock(name='c.client')
+        self.backend.one_client.kv.delete.return_value = True
+        assert self.backend.delete('Key') is True
+
     def test_index_bytes_key(self):
         key = 'test-consul-2'
         assert self.backend._key_to_consul_key(key) == key

